### PR TITLE
Locking submissions

### PR DESCRIPTION
**Pull request by almogcarmi**
- Added validation for every Moodle API calling.
- Added logging in some of the places.
- Added assignments retrieve logic by name and by section.
- Added a sections.py file.

### CODE REVIEW

File: moodler/moodle.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
 from moodler.config import STUDENTS_TO_IGNORE
 from moodler.download import download_file, download_submission
 from moodler.feedbacks import feedbacks
-from moodler.students import get_students, core_course_get_contents
+from moodler.students import get_students
+from moodler.sections import core_course_get_contents
```
Comment (maxandron) [inline moodler/moodle.py:10]: Unused imports?

File: moodler/consts.py [hunk 2]
```diff
@@ -3,3 +3,23 @@
...
+def validate_response(function_name, response):
+    """
+    Function that validates whether the response received is a valid response or
+    is it an exception.
+    """
+    if 'exception' in response:
+        raise MoodleAPIException("Moodle API call for function '{}' returned "
+                                 "an exception response with the following "
+                                 "message: {}".format(function_name,
+                                                      response['message']))
```
Comment (maxandron) [inline moodler/consts.py]: Maybe this should be in moodle.py?

File: moodler/assignment.py [hunk 3]
```diff
@@ -60,6 +60,60 @@ def ungraded(self):
...
+def mod_assign_lock_submissions(assignment_id, user_ids):
+    """
+    Locks submissions for a specific assignments for a specific user(s).
+    """
+    url = REQUEST_FORMAT.format('mod_assign_lock_submissions')
+
+    url += '&assignmentid={}'.format(assignment_id)
+
+    for i, user_id in enumerate(user_ids):
+        url += '&userids[{}]={}'.format(i, user_id)
```
Comment (maxandron) [inline moodler/assignment.py]: This code is duplicated all over. Maybe can create a helper function?

File: moodler/assignment.py [hunk 4]
```diff
@@ -102,12 +160,64 @@ def get_assignments(course_id, assignment_ids_to_get=None):
...
+        assignments.append(Assignment(assignment,
+                                      submissions.get(assignment['id'], []),
+                                      grades.get(assignment['id'], [])))
+
+        if assignments_not_found is not None:
+            assignments_not_found.remove(assignment['name'])
+
+    if assignments_not_found:
+        logger.error("Could not find the following exercises for the trainer: "
+                     "%s", assignments_not_found)
```
Comment (maxandron) [inline moodler/assignment.py]: Where is this function used? in dart_training?

File: moodler/students.py [hunk 5]
```diff
@@ -1,64 +1,69 @@
...
     """
-    enrolled_students = {}
+    students_ids = []
+
     for enrolled in core_enrol_get_enrolled_users(course_id):
-        if enrolled['roles'][0]['shortname'] == 'student':
-            enrolled_students[enrolled['id']] = enrolled['fullname']
-    return enrolled_students
+        if enrolled['roles'][0]['shortname'] != 'student':
+            continue
```
Comment (maxandron) [inline moodler/students.py]: Isn't it just get_students with an extra parameter - can't these two functions be merged?

File: moodler/sections.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        if course_prefix is not None:
+            if ((course_prefix not in course['shortname']) and
+                    (course_prefix not in course['fullname'])):
+                continue
+
+        courses_list.append(Course(course['id'],
+                                   course['shortname'],
+                                   course['fullname']))
+
+    return courses_list
```
Comment (maxandron) [inline moodler/sections.py:85]: Maybe just use the `filter()` function here?

**Review by maxandron — CHANGES_REQUESTED** [review]

File: moodler/moodle.py (hunk 1 above)
Comment (almogcarmi) [reply moodler/moodle.py:10]: All of them are used in this file...

File: moodler/assignment.py (hunk 4 above)
Comment (almogcarmi) [reply moodler/assignment.py]: Yes this function is used in dart_training.
Also combined with the get_assignments function you can see it.

File: moodler/consts.py (hunk 2 above)
Comment (almogcarmi) [reply moodler/consts.py]: Moved this to another file called moodle_api for any generic utility functions used for moodle API calls.

File: moodler/assignment.py (hunk 3 above)
Comment (almogcarmi) [reply moodler/assignment.py]: Done :)
Take a look at the function call_moodle_api.
Notice the function list_students() in the students.py file does not use this utility function but calls the requests.get directly. If you read the way you pass core_user_get_users function parameters, you will understand why this is the only function that I left as is.

File: moodler/students.py (hunk 5 above)
Comment (almogcarmi) [reply moodler/students.py]: Fixed

File: moodler/sections.py (hunk 6 above)
Comment (almogcarmi) [reply moodler/sections.py:85]: The reason I do not use filter is that I need to build the course objects along the way so I have to iterate the list of course anyways...

**Review by maxandron — APPROVED** [review]

**Pull request merged**